packages/indexer/src/sinks/sqlite.ts

```diff
@@ -0,0 +1,111 @@
+import type { Cursor } from "@apibara/protocol";
+import { open, type Database, type ISqlite } from "sqlite";
+import { Sink, type SinkWriteArgs } from "../sink";
+
+export type SqliteArgs = ISqlite.Config;
+
+export type SqliteSinkOptions = {
+  /**
+   * The name of the table where data will be inserted.
+   */
+  tableName: string;
+  /**
+   * An optional column name used to store the cursor value. If specified,
+   * the value of this column must match the `endCursor.orderKey` for each row.
+   */
+  cursorColumn?: string;
+  /**
+   * An optional configuration to handle conflicts during data insertion.
+   * - `on`: The column name on which conflicts are detected.
+   * - `update`: An array of column names to be updated if a conflict occurs.
+   */
+  onConflict?: { on: string; update: string[] };
+};
+
+class SqliteSink<TData = unknown> extends Sink<TData> {
+  private _config: SqliteSinkOptions;
+  private _db: Database;
+
+  constructor(db: Database, config: SqliteSinkOptions) {
+    super();
+    this._config = config;
+    this._db = db;
+  }
+
+  async write({ data, endCursor }: SinkWriteArgs<TData>) {
+    this.emit("write", { data });
+
+    const { cursorColumn } = this._config;
+
+    data = this.processCursorColumn(data, cursorColumn, endCursor);
+
+    await this.insertJsonArray(data);
+
+    this.emit("flush");
+  }
+
+  private async insertJsonArray(data: TData) {
+    if (!Array.isArray(data) || data === null) {
+      throw new Error("Data is not an array or is null");
+    }
+
+    if (data.length === 0) return;
+
+    // Get columns from the first row of the object array
+    const columns = Object.keys(data[0]);
+    const columnNames = columns.join(", ");
+    const placeholders = columns.map(() => "?").join(", ");
+
+    // Handle onConflict option
+    const conflictClause = this.buildConflictClause();
+
+    // Build the SQL insert statement with multiple rows
+    const insertSQL = `INSERT INTO ${this._config.tableName} (${columnNames}) VALUES `;
+    const valuePlaceholders = data.map(() => `(${placeholders})`).join(", ");
+    const statement = insertSQL + valuePlaceholders + conflictClause;
+
+    // Prepare and execute the SQL statement
+    const values = data.flatMap((row) => columns.map((col) => row[col]));
+
+    await this._db.run(statement, values);
+  }
+
+  private processCursorColumn(
+    data: TData,
+    cursorColumn?: string,
+    endCursor?: Cursor,
+  ): TData {
+    if (!Array.isArray(data) || data === null) {
```

Comment:
Since the argument to `write` is now `TData[]` there is no need to check it's an array anymore.